traversal/GraphTraversal.java

```diff
@@ -130,154 +131,108 @@ public Type() {
-            return filter;
-        }
-
-        @Override
-        public void filter(Set<? extends Identifier.Variable.Retrievable> filter) {
-            assert iterate(filter).noneMatch(Identifier::isLabel);
-            this.filter.addAll(filter);
+            return new CombinationFinder(graphMgr, CombinationProcedure.create(structure), modifiers.filter(), concreteVarIds).combination();
         }
     }
 
     public static class Thing extends GraphTraversal {
 
         private Planner planner;
         private TraversalCache cache;
-        private boolean modifiable;
 
         public Thing() {
             super();
-            modifiable = true;
         }
 
         public void initialise(TraversalCache cache) {
             assert planner == null;
             this.cache = cache;
-            planner = this.cache.getPlanner(structure, Planner::create);
+            planner = this.cache.getPlanner(structure, modifiers, sm -> Planner.create(sm.first(), sm.second()));
         }
 
         @Override
         FunctionalIterator<VertexMap> permutationIterator(GraphManager graphMgr) {
             assert planner != null && cache != null;
-            FunctionalIterator<VertexMap> iter = permutationIterator(graphMgr, planner, false, filter());
-            cache.mayUpdatePlanner(structure, planner);
+            FunctionalIterator<VertexMap> iter = permutationIterator(graphMgr, planner, false);
+            cache.mayUpdatePlanner(structure, modifiers, planner);
             return iter;
         }
 
         FunctionalProducer<VertexMap> permutationProducer(GraphManager graphMgr,
                                                           int parallelisation) {
             assert planner != null && cache != null;
             planner.tryOptimise(graphMgr, false);
-            FunctionalProducer<VertexMap> producer = planner.procedure().producer(graphMgr, parameters, filter(), parallelisation);
-            cache.mayUpdatePlanner(structure, planner);
+            FunctionalProducer<VertexMap> producer = planner.procedure().producer(graphMgr, parameters, modifiers, parallelisation);
+            cache.mayUpdatePlanner(structure, modifiers, planner);
             return producer;
         }
 
-        // TODO: We should not dynamically calculate properties like this, and then guard against 'modifiable'.
-        //       We should introduce a "builder pattern" to Traversal, such that users of this library will build
-        //       traversals with Traversal.Builder, and call .build() in the end to produce a final Object.
-        @Override
-        Set<Identifier.Variable.Retrievable> filter() {
-            if (filter.isEmpty()) {
-                modifiable = false;
-                iterate(structure.vertices()).filter(v -> v.id().isRetrievable()).map(v -> v.id().asVariable().asRetrievable())
-                        .toSet(filter);
-            }
-            return filter;
-        }
```

Comment:
creating the default filter has moved up to the user, eg. `Conjunction.traversal(filter, sorting)`, which sets the filter to be all retrievable variables if the user filter is empty.